src/compiler/compile/render_dom/wrappers/Element/index.ts

```diff
@@ -1087,19 +1103,23 @@ export default class ElementWrapper extends Wrapper {
 
 			block.chunks.hydrate.push(updater);
 
-			const dependencies = expression.dynamic_dependencies();
+			// Assume that style has changed through the spread attribute
 			if (has_spread) {
 				block.chunks.update.push(updater);
-			} else if (dependencies.length > 0) {
-				if (should_cache) {
-					block.chunks.update.push(b`
-							if (${block.renderer.dirty(dependencies)} && (${cached_snippet} !== (${cached_snippet} = ${snippet}))) {
-								${updater}
-							}
-					`);
-				} else {
+			} else {
+				const dependencies = [
+					...this.dynamic_style_dependencies,
+					...expression.dynamic_dependencies()
+				];
+
+				if (dependencies.length > 0) {
+					const is_dirty = block.renderer.dirty(dependencies);
+					const condition = should_cache
+						? x`${is_dirty} && (${cached_snippet} !== (${cached_snippet} = ${snippet}))`
+						: is_dirty;
+
 					block.chunks.update.push(b`
-						if (${block.renderer.dirty(dependencies)}) {
+						if (${condition}) {
 							${updater}
```

Comment:
i feel the style directive and style attribute update logic can be optimised,

rather than having disable the caching when there's a style dependencies, we can use another flag to determine whether to call the updater:

```js
p(ctx, [dirty]) {
  let is_style_updated = false;
  if (dirty & /* style dependencies */ 2 && ...) {
    // style update
    attr(button, "style", button_style_value);
    is_style_updated = true;
  }

  if (is_style_updated || (dirty & /* style directive dependencies */ 1 && style_color !== (style_color = `...`))) {
    set_style(button, "color", style_color, false);
  }
}
```

and i also think that if there's no dependencies on the style directives, eg:

```svelte
<button
  style={pressed ? 'font-size: 10px' : null}
  style:color="blue" />
```

we can even simplified the logic further to:

```js
p(ctx, [dirty]) {
  if (dirty & /* style dependencies */ 2 && ...) {
    // style update
    attr(button, "style", button_style_value);
    set_style(button, "color", style_color, false);
  }

  // no need extra block for style directives
}
```

the code would be more optimised, though this would require more code changes over here

Reply:
```js
let is_style_updated = dirty & /*style deps*/ 1 && ...;
if (is_style_updated) {
	attr(.....)
}
```
might be better.

I was thinking of something similar at the time but then the `attr -> set_style` optimization in StyleAttributeWrapper got in the way iirc.

Reply:
well, we can single out the `StyleAttributeWrapper` from `this.attributes` and process the `StyleAttributeWrapper` with other style directives